src/App.tsx

```diff
@@ -15,48 +21,112 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+// eslint-disable-next-line max-len
+const stableSort = (array: string[], compareFunction: (a: string, b: string) => number) => {
+  return array
+    .map((item, index) => ({ item, index }))
+    .sort((a, b) => compareFunction(a.item, b.item) || a.index - b.index)
+    .map(({ item }) => item);
+};
+
 export const App: React.FC = () => {
+  const [goods, setGoods] = useState<string[]>(initialGoods);
+  const [sortType, setSortType] = useState<SortType>(SortType.Default);
+  const [ascending, setAscending] = useState<boolean>(true);
+  // eslint-disable-next-line max-len
+  const [originalOrderReversed, setOriginalOrderReversed] = useState<boolean>(false);
+
+  const handleSort = (type: SortType) => {
+    let sortedGoods = [...goods];
+
+    if (type === SortType.Alphabetically) {
+      sortedGoods = stableSort([...goods], (a, b) => a.localeCompare(b));
+    } else if (type === SortType.Length) {
+      sortedGoods = stableSort(
+        [...goods],
+        (a, b) => a.length - b.length || goods.indexOf(a) - goods.indexOf(b),
```

Comment:
It's enough to do sort like this `copyGoods.sort((a, b) => a.length - b.length);`.